packages/core/src/transport/flushController.spec.ts

```diff
@@ -78,6 +81,33 @@ describe('flushController', () => {
     })
   })
 
+  describe('session expire', () => {
+    it('notifies when the session expires', () => {
+      flushController.notifyBeforeAddMessage(SMALL_MESSAGE_BYTE_COUNT)
+      flushController.notifyAfterAddMessage()
+      sessionExpireObservable.notify()
+      expect(flushSpy).toHaveBeenCalled()
+    })
+
+    it('flush reason should be "session_expire"', () => {
+      flushController.notifyBeforeAddMessage(SMALL_MESSAGE_BYTE_COUNT)
+      flushController.notifyAfterAddMessage()
+      sessionExpireObservable.notify()
+      expect(flushSpy.calls.first().args[0].reason).toBe('session_expire')
+    })
+
+    it('does not notify if no message was added', () => {
+      sessionExpireObservable.notify()
+      expect(flushSpy).not.toHaveBeenCalled()
+    })
+
+    it('notifies when the session expires even if no message have been fully added yet', () => {
+      flushController.notifyBeforeAddMessage(SMALL_MESSAGE_BYTE_COUNT)
+      sessionExpireObservable.notify()
+      expect(flushSpy).toHaveBeenCalled()
+    })
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏I am wondering if we could not have a chance to have an incorrect state in this case 🤔

| operation | flushController | batch/segment |
|----------| ----------------|----------------|
| notifyBeforeAddMessage() | messageCount = 1 | bufferSize = 0 |
| expire flush | messageCount = 0 | bufferSize = 0 |
| addMessage | messageCount = 0 | bufferSize = 1 |
| notifyAfterAddMessage() | messageCount = 0 | bufferSize = 1 |

In this case a subsequent flush without other messages added would incorrectly not trigger the batch/segment flush, right?

Reply:
This is true if `expire flush` can happen between `notifyBeforeAddMessage` and `addMessage`. In practice, `notifyBeforeAddMessage` and `addMessage` are happening consecutively and nothing can happen between the two.

This invariant cannot be guaranteed with the current design, but I thought that it was fine for the sake of simplicity.
I'll write a bit of documentation to clarify.

Reply:
> In practice, notifyBeforeAddMessage and addMessage are happening consecutively and nothing can happen between the two.

True, I am wondering if it will still be the case when we will plug the flush controller to segment collection 🤔
Could it worth to do or plan something?

Reply:
Yes it will still be true for segment collection.

The idea is to call `notifyBeforeAddMessage` right before sending the message/record to the webworker, so it cannot be flushed between the two. Only `notifyAfterAddMessage` will be called asynchronously, when we know the message compressed size.